dijet/tasks/correlated_fit.py

```diff
@@ -0,0 +1,89 @@
+# coding: utf-8
+
+import law
+from columnflow.util import maybe_import
+
+from dijet.tasks.util import chi2
+
+sc = maybe_import("scipy.optimize")
+nd = maybe_import("numdifftools")
+np = maybe_import("numpy")
+
+class CorrelatedFit():
+    """
+    Task to calculated correlated fit to widths.
+
+    ## Covariance matrix 
+    [1] Based on K. Goebel dissertation ch. A.3.
+    https://www.physik.uni-hamburg.de/en/iexp/gruppe-haller/scientific-output/documents/thesis-kristin-goebel.pdf
+
+    Example code here
+    https://github.com/UHH2/DiJetJERC/blob/ff98eebbd44931beb016c36327ab174fdf11a83f/
+    JERSF_Analysis/JER/wide_eta_binning/functions.C#L558
+    func. createCov()
+    """
+
+    def createCov(self, widths, widths_err):
+        widths_err2 = pow(widths_err, 2)  # square each element
+
+        # 3x3 matrix with the smaller error of indices i&j (Check [1])
```

Comment:
Is it the smaller or larger error? `np.maximum.outer` seems to suggest the latter.

Reply:
In this line it is more important to get the structer as below to what the 1D array is transformed to. This is checked to be true by this operation.

Reply:
Not sure I follow, the values in `matrix_err` after this line are directly used in the covariance matrix calculation, no? Maybe we could cross-check again with what is actually done in the reference you provided.

Reply:
I checked the implementation against the reference provided and there are a few mistakes here. First, `np.minimum.outer` should be calculated based on the indices, but here the width values themselves are used, meaning that sometimes the wrong width_err2 is chosen. Second, the ratio of events appears with a square here, but it is not squared in the original reference.

Another more minor issue is that the event counts are inferred from the errors, which themselves are calculated elsewhere based on the event counts. We should avoid transforming back and forth like this and instead provide the event counts directly to the `create_cov` function.

I therefore suggest to implement this differently. Here is piece of code that should do what the reference describes:
```python
@staticmethod
def create_cov(widths, n):
    """
    Calculate a covariance matrix for a sequence of distribution widths
    correspondin to inclusive alpha bins. The input values must be in
    order of ascending alpha.

    widths: array of widths of the asymmetry distributions in the inclusive alpha bins
    n: array of event counts in the inclusive alpha bins
    """
    # square of the uncertainty on the width from normal approximation
    widths_err2 = widths**2 / (2 * n)

    # NxN matrices mapping pair of values (a_i, a_j) to that
    # corresponding to the lower (higher) index,
    # i.e. entry at (i, j) = min(i, j) or max(i, j)
    idx = np.arange(len(widths))
    min_idx_matrix = np.minimum.outer(idx, idx)
    max_idx_matrix = np.maximum.outer(idx, idx)

    # obtain ratios of widths and event counts
    # for larger over smaller index
    width_matrix = widths[min_idx_matrix] / widths[max_idx_matrix]
    n_matrix = n[min_idx_matrix] / n[max_idx_matrix]

    # calculate covariance and return
    cov_mat = width_matrix * n_matrix * widths_err2
    return np.nan_to_num(cov_mat)
```